### PR TITLE
Fix JSON string escape for profiling data output

### LINKED ISSUE #18361: Google trace format (JSON) data not escaped properly
There is currently no escaping being done in the code that outputs the JSON output from https://github.com/coq/coq/pull/17702:
https://github.com/coq/coq/blob/1788628756921f2b3684023ec7f0628a53ac0b26/lib/newProfile.ml#L23

As a consequence, commands that contain, say, a `\` as part of a notation are printed unescaped, which makes some JSON parsers (at least that of `Yojson`, with its default config) choke.

Potentially, changing the format from `"\"%s\""` to `"%S"` could help, but would probably not be entirely correct. Maybe a proper JSON printer should be used? CC @SkySkimmer

### PR DESCRIPTION
**Pull request by rlepigre**
Fixes https://github.com/coq/coq/issues/18361.

~~I have not done much testing, and I'm not sure my interpretation of https://www.ietf.org/rfc/rfc4627.txt is entirely correct, since it is phrased quite ambiguously as to whether `/` should be escaped or not.~~

Edit: escaping exactly as Yojson does.

### CODE REVIEW

**Comment by JasonGross** [comment]
> I have not done much testing, and I'm not sure my interpretation of https://www.ietf.org/rfc/rfc4627.txt is entirely correct, since it is phrased quite ambiguously as to whether `/` should be escaped or not.

Are you looking at section 2.5, or some other section?  My reading of the `unescaped` line of
```
         string = quotation-mark *char quotation-mark

         char = unescaped /
                escape (
                    %x22 /          ; "    quotation mark  U+0022
                    %x5C /          ; \    reverse solidus U+005C
                    %x2F /          ; /    solidus         U+002F
                    %x62 /          ; b    backspace       U+0008
                    %x66 /          ; f    form feed       U+000C
                    %x6E /          ; n    line feed       U+000A
                    %x72 /          ; r    carriage return U+000D
                    %x74 /          ; t    tab             U+0009
                    %x75 4HEXDIG )  ; uXXXX                U+XXXX

         escape = %x5C              ; \

         quotation-mark = %x22      ; "

         unescaped = %x20-21 / %x23-5B / %x5D-10FFFF
```
is that we only actually need to escape `\x22` (`"`) and `\x5C` (`\`).

**Comment by rlepigre** [comment]
Yeah, I was referring to the start of section 2.5:
> 2.5.  Strings
>
>   The representation of strings is similar to conventions used in the C
>   family of programming languages.  A string begins and ends with
>   quotation marks.  All Unicode characters may be placed within the
>   quotation marks except for the characters that must be escaped:
>   quotation mark, reverse solidus, and the control characters (U+0000
>   through U+001F).

If you just stick to that bit, it only really takes about `"`, `\` and the control characters from `0x00` to `0x1F`.

**Comment by ejgallego** [comment]
I agree with @JasonGross assesment, `/` _may_ appear escaped, so `\/` needs to be parsed as `/`, however note how `0x2F` is in the `unescaped` set, so it can appear unescaped too.

File: lib/newProfile.ml [hunk 1]
```diff
@@ -16,11 +16,27 @@ module MiniJson = struct
...
+    let buf = Buffer.create (String.length s) in
+    let encode c =
+      match c with
+      | '\b' -> Buffer.add_string buf "\\b"
+      | '\t' -> Buffer.add_string buf "\\t"
+      | '\n' -> Buffer.add_string buf "\\n"
+      | '\012' -> Buffer.add_string buf "\\f"
+      | '\r' -> Buffer.add_string buf "\\r"
+      | '\x00'..'\x21' | 0x7F ->
+          Buffer.add_string buf (Printf.sprintf "\\u%04x" (Char.code c))
```
Comment (ejgallego) [inline lib/newProfile.ml:30]: It seems to me that this range should stop at `\x1F`. Escaping ` ` (space) and `!` (bang) seems not to be necessary.

File: lib/newProfile.ml [hunk 2]
```diff
@@ -16,11 +16,27 @@ module MiniJson = struct
...
+      | '"' -> Buffer.add_string buf "\\\""
+      | '\\' -> Buffer.add_string buf "\\\\"
+      | c -> Buffer.add_char buf c
```
Comment (JasonGross) [inline lib/newProfile.ml:33]: Do we want to go with the minimal quoting?
```suggestion
      | '\x00'..'\x1F' ->
          Buffer.add_string buf (Printf.sprintf "\\u%04x" (Char.code c))
      | '"' -> Buffer.add_string buf "\\\""
      | '\\' -> Buffer.add_string buf "\\\\"
      | c -> Buffer.add_char buf c
```

Comment (ejgallego) [reply lib/newProfile.ml:33]: Should work, tho I'd suggest to follow the yojson set for the reasons stated in the discussion.

**Comment by rlepigre** [comment]
OK, so in the end I just went with the same escaping scheme as Yojson, which also adds `0x7F` to the set of escaped control characters for some reason.

File: lib/newProfile.ml [hunk 3]
```diff
@@ -16,11 +16,27 @@ module MiniJson = struct
     | `List of t list
   ]
 
+  (* https://www.ietf.org/rfc/rfc4627.txt *)
```
Comment (ejgallego) [inline lib/newProfile.ml:19]: You may want to add a link to the yojson version, and maybe reorder the match to follow it?

This way we can quickly check if we have deviated.

File: lib/newProfile.ml (hunk 1 above)
Comment (rlepigre) [reply lib/newProfile.ml:34]: Good catch.

**Comment by ejgallego** [comment]
@SkySkimmer is the milestone OK for you?

**Review by ejgallego — APPROVED** [review]
Thanks!

File: lib/newProfile.ml [hunk 4]
```diff
@@ -16,11 +16,29 @@ module MiniJson = struct
...
+      match c with
+      | '"' -> Buffer.add_string buf "\\\""
+      | '\\' -> Buffer.add_string buf "\\\\"
+      | '\b' -> Buffer.add_string buf "\\b"
+      | '\012' -> Buffer.add_string buf "\\f"
+      | '\n' -> Buffer.add_string buf "\\n"
+      | '\r' -> Buffer.add_string buf "\\r"
+      | '\t' -> Buffer.add_string buf "\\t"
+      | '\x00'..'\x1F'
+      | 0x7F ->
```
[the remark this answers is not in the document]
Comment (rlepigre) [reply lib/newProfile.ml:33]: Yep, it seems like I can't type today.

Comment (rlepigre) [inline lib/newProfile.ml:33]:
```suggestion
      | '\x7F' ->
```

**Comment by ejgallego** [comment]
The more time you spent looking at the JSON spec the more your typing abilities decrease XD

Do you mind squashing and pushing again? That will launch the full CI.

**Comment by rlepigre** [comment]
> Do you mind squashing and pushing again? That will launch the full CI.

Done.

**Comment by ejgallego** [comment]
@coqbot run full CI

**Pull request merged**